sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,316 @@
+import numpy as np
+
+from ..base import BaseEstimator, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+from ..exceptions import ConvergenceWarning
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if not hasattr(estimator, "predict_proba"):
+        name = type(estimator).__name__
+        msg = "base_classifier ({}) should implement predict_proba!"
+        msg = msg.format(name)
+        raise ValueError(msg)
+
+
+class SelfTrainingClassifier(BaseEstimator):
+    """Self-training classifier
+
+    This class allows a given supervised classifier to function as a
+    semi-supervised classifier, allowing it to learn from unlabeled data. It
+    does this by iteratively predicting labels for the unlabeled data.
+
+    Read more in the :ref:`User Guide <self_training>`.
+
+    Parameters
+    ----------
+    base_classifier : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``base_classifier_`` attribute.
+
+    threshold : float, optional (default=0.75)
+        The decision threshold. If the ``base_classifier`` makes a prediction
+        with a ``predict_proba`` above this threshold, it will be added to the
+        labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : int or ``None``, optional (default=10)
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0. If it is ``None``, the classifier will continue to predict labels
+        until all unlabeled samples have been labeled. In this case, be aware
+        that the fit may never terminate.
+
+    n_iter_no_change: int, optional (default=3)
```

Comment:
What are your thoughts on using a fractional change?

Reply:
I'm not opposed to adding it, but both `SGDClassifier` and `GradientBoostingClassifier` also only have `int`, so i'm not sure if it's something common in practice.

Reply:
Since `max_iter` is set to `10` by default, I think this being an integer is okay. I would leave it as is.